tensorflow/python/keras/engine/training.py

```diff
@@ -1084,27 +1084,41 @@ def fit(self,
       # Handle fault-tolerance for multi-worker.
       # TODO(omalleyt): Fix the ordering issues that mean this has to
       # happen after `callbacks.on_train_begin`.
-      data_handler._initial_epoch = (  # pylint: disable=protected-access
-          self._maybe_load_initial_epoch_from_ckpt(initial_epoch))
-      logs = None
+      initial_epoch = self._maybe_load_initial_epoch_from_ckpt(initial_epoch)
+      data_handler._initial_epoch = initial_epoch  # pylint: disable=protected-access
+      run_compiled = (not self.run_eagerly and
+                      not callbacks._should_call_train_batch_hooks and  # pylint: disable=protected-access
+                      self._steps_per_execution.numpy().item() == 1)
+      logs, epoch_train_fn = None, None
       for epoch, iterator in data_handler.enumerate_epochs():
         self.reset_metrics()
         callbacks.on_epoch_begin(epoch)
-        with data_handler.catch_stop_iteration():
-          for step in data_handler.steps():
-            with trace.Trace(
-                'train',
-                epoch_num=epoch,
-                step_num=step,
-                batch_size=batch_size,
-                _r=1):
-              callbacks.on_train_batch_begin(step)
-              tmp_logs = self.train_function(iterator)
-              if data_handler.should_sync:
-                context.async_wait()
-              logs = tmp_logs  # No error, now safe to assign to logs.
-              end_step = step + data_handler.step_increment
-              callbacks.on_train_batch_end(end_step, logs)
+        if (run_compiled and data_handler.inferred_steps is not None and
+            epoch - initial_epoch > 0):
```

Comment:
We only use a `tf.function` for the second epoch as we need `logs` to be defined in order to prevent us needing to run the first train step outside the loop similar to:
https://github.com/tensorflow/tensorflow/blob/3a3d3591923501aa6fb03ef4da1d64400ad209d8/tensorflow/python/keras/engine/training.py#L813-L818

If I understand https://github.com/tensorflow/tensorflow/pull/40881#issuecomment-662731863 correctly this would otherwise lead to an increased graph size which can become problematic for larger networks.

In theory we could also only run the first training step outside of the `tf.function`, but I think this would make the code a bit more complicate and won't increase performance by much.